torch/serialization.py

```diff
@@ -399,6 +405,68 @@ def persistent_id(obj):
         serialized_storages[key]._write_file(f, _should_read_directly(f))
 
 
+def _save(obj, zip_file, pickle_module, pickle_protocol):
+    serialized_storages = {}
+
+    def persistent_id(obj):
+        # FIXME: the docs say that persistent_id should only return a string
+        # but torch store returns tuples. This works only in the binary protocol
+        # see
+        # https://docs.python.org/2/library/pickle.html#pickling-and-unpickling-external-objects
+        # https://github.com/python/cpython/blob/master/Lib/pickle.py#L527-L537
+        if torch.is_storage(obj):
+            storage_type = normalize_storage_type(type(obj))
+            # Offset is always 0, but we keep it for backwards compatibility
+            # with the old serialization format (which supported storage views)
+            offset = 0
+            obj_key = str(obj._cdata)
+            location = location_tag(obj)
+            serialized_storages[obj_key] = obj
+
+            return ('storage',
+                    storage_type,
+                    obj_key,
+                    location,
+                    obj.size())
+        return None
+
+    sys_info = {
+        "little_endian": sys.byteorder == 'little',
+        "type_sizes": {
+            "short": SHORT_SIZE,
+            "int": INT_SIZE,
+            "long": LONG_SIZE,
+        }
+    }
+
+    metadata = {
+        "magic_number": MAGIC_NUMBER,
+        "protocol_version": ZIPFILE_PROTOCOL_VERSION,
+        "sys_info": sys_info
+    }
+
+    # Write system metadata
+    metadata_buf = io.BytesIO()
+    pickle_module.dump(metadata, metadata_buf)
+    metadata_value = metadata_buf.getvalue()
+    zip_file.write_record('metadata.pkl', metadata_value, len(metadata_value))
```

Comment:
Not a full review yet, but we need to be extremely careful deciding what extra information goes in the zip file. This stuff causes infinite tech debt, and I do not see a good justificaiton for most of the things in the metadata pickle. Also, naming things metadata is not a great idea. It is better to name it what it actually contains. 

* I don't see type_sizes being used, so why are they being saved? 
* Our zip files already contain version numbers (see inline container), please do not add another.
* The presence of the right files in the zip file should be sufficient to establish this as the right kind of archive, having an additional magic number beyond that feels unnecessary, and we have to do error checking on the individual files anyway. At the very least, magic numbers need to be consistent across jit/non-jit or we gain nothing by moving eager to JIT.
* I don't understand PROTOCOL_VERSION here? That was the version number for the old  format, but it simply can't exist at the same byte location in the new format. What is its purpose here?

Reply:
(also, to clarify, I don't think we need a file `metadata.pkl` at all when we boil down what is necessary).

Reply:
Those are mostly there to maintain the same information as the current serialization format. I'm fine getting rid of them since they're not really public facing